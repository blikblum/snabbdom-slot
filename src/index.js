/**
 * snabbdom-slot (https://github.com/blikblum/snabbdom-slot)
 *
 * Copyright © 2018 Luiz Américo Pereira Câmara. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

export { withSlot } from './withslot'
