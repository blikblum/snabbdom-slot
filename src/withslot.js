/* eslint-disable no-cond-assign */
function buildSlotMap (vnode, map) {
  let child
  if (vnode.children) {
    for (let i = 0; i < vnode.children.length; i++) {
      child = vnode.children[i]
      if (child.sel === 'slot') {
        map[getSlotName(child)] = {
          node: child,
          parent: vnode,
          index: i
        }
      }
      buildSlotMap(child, map)
    }
  }
}

function getSlotName (vnode) {
  return (vnode.data.props && vnode.data.props.name) || ''
}

function redistributeNodes (children, map) {
  const withoutSlot = []
  const dupes = {}
  let slotData
  children.forEach((child) => {
    const slot = child.data.props && child.data.props.slot
    if (slot) {
      if (slotData = map[slot]) {
        if (!dupes[slot]) {
          slotData.node.children = [child]
          slotData.node.text = undefined
          dupes[slot] = true
        } else {
          slotData.node.children.push(child)
        }
      }
    } else {
      withoutSlot.push(child)
    }
  })
  if ((slotData = map['']) && withoutSlot.length) {
    slotData.node.children = withoutSlot
    slotData.node.text = undefined
  }
}

export const withSlot = (renderFn) => {
  return function (props, children) {
    const containerTree = renderFn(props, children)
    const slotMap = Object.create(null)
    if (containerTree.sel === 'slot') {
      slotMap[getSlotName(containerTree)] = {
        node: containerTree
      }
    }
    if (!Array.isArray(children)) children = [children]
    buildSlotMap(containerTree, slotMap)
    redistributeNodes(children, slotMap)
    return containerTree
  }
}
