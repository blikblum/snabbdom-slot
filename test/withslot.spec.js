/* global describe, it, expect */

import { withSlot } from '../src/withslot'
import render from './render-tree'
import Snabbdom from 'snabbdom-pragma'

describe('when the component has no slot', () => {
  const NoSlot = withSlot(() => {
    return (
      <div>
        <div>No slot</div>
      </div>
    )
  })
  it('children is ignored', () => {
    const tree = (
      <NoSlot>
        <div>Child element</div>
      </NoSlot>
    )
    expect(render(tree)).toMatchSnapshot()
  })
})

describe('when the component has a main slot', () => {
  const MainSlot = withSlot(() => {
    return (
      <div>
        <slot>Default content</slot>
      </div>
    )
  })

  describe('and no children is passed', () => {
    it('should show default content', () => {
      const tree = (
        <MainSlot/>
      )
      expect(render(tree)).toMatchSnapshot()
    })
  })

  describe('and children is passed', () => {
    it('should show children content', () => {
      const tree = (
        <MainSlot>
          <h1>Child element</h1>
          <div>Other child element</div>
        </MainSlot>
      )
      expect(render(tree)).toMatchSnapshot()
    })
  })
})

describe('when the component has a named slot', () => {
  const NamedSlot = withSlot(() => {
    return (
      <div>
        <slot name='myslot'>Default Content</slot>
      </div>
    )
  })

  describe('and no children is passed', () => {
    it('should show default content', () => {
      const tree = (
        <NamedSlot/>
      )
      expect(render(tree)).toMatchSnapshot()
    })
  })

  describe('and children without matching slot prop is passed', () => {
    it('should show default content', () => {
      const tree = (
        <NamedSlot>
          <h1 slot='myotherslot'>Child element</h1>
        </NamedSlot>
      )
      expect(render(tree)).toMatchSnapshot()
    })
  })

  describe('and a child with matching slot prop is passed', () => {
    it('should show child element in correspondent slot element', () => {
      const tree = (
        <NamedSlot>
          <h1 slot='myslot'>Child element</h1>
        </NamedSlot>
      )
      expect(render(tree)).toMatchSnapshot()
    })
  })
})
