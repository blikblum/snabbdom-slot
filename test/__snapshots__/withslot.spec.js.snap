// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the component has a main slot and children is passed should show children content 1`] = `
"<div>
    <slot>
        <h1>Child element</h1>
        <div>Other child element</div>
    </slot>
</div>"
`;

exports[`when the component has a main slot and no children is passed should show default content 1`] = `
"<div>
    <slot>Default content</slot>
</div>"
`;

exports[`when the component has a named slot and a child with matching slot prop is passed should show child element in correspondent slot element 1`] = `
"<div>
    <slot name=\\"myslot\\">
        <h1 slot=\\"myslot\\">Child element</h1>
    </slot>
</div>"
`;

exports[`when the component has a named slot and children without matching slot prop is passed should show default content 1`] = `
"<div>
    <slot name=\\"myslot\\">Default Content</slot>
</div>"
`;

exports[`when the component has a named slot and no children is passed should show default content 1`] = `
"<div>
    <slot name=\\"myslot\\">Default Content</slot>
</div>"
`;

exports[`when the component has no slot children is ignored 1`] = `
"<div>
    <div>No slot</div>
</div>"
`;
